core/corehttp/gateway_handler.go

```diff
@@ -781,3 +808,28 @@ func preferred404Filename(acceptHeaders []string) (string, string, error) {
 
 	return "", "", fmt.Errorf("there is no 404 file for the requested content types")
 }
+
+// Attempt to fix redundant /ipfs/ namespace as long as resulting
+// 'intended' path is valid.  This is in case gremlins were tickled
+// wrong way and user ended up at /ipfs/ipfs/{cid} or /ipfs/ipns/{id}
+// like in bafybeien3m7mdn6imm425vc2s22erzyhbvk5n3ofzgikkhmdkh5cuqbpbq :^))
+func fixupSuperfluousNamespace(w http.ResponseWriter, r *http.Request, urlPath string) bool {
+	if !(strings.HasPrefix(urlPath, "/ipfs/ipfs/") || strings.HasPrefix(urlPath, "/ipfs/ipns/")) {
+		return false // not a superfluous namespace
+	}
+	intendedPath := ipath.New(strings.TrimPrefix(urlPath, "/ipfs"))
+	if err := intendedPath.IsValid(); err != nil {
+		return false // not a valid path
+	}
+	intendedURL := strings.Replace(r.URL.String(), urlPath, intendedPath.String(), 1)
```

Comment:
This makes me a little nervous. Can't this be mounted at some `/ipfs` prefix? Probably fine, just a bit funky.

Reply:
But I'm fine with it as-is.

Reply:
@Stebalien good catch. Been a while, iirc I did this  to  preserve `?query`  params, but it is not safe to write it into HTML.

Simplified this in https://github.com/ipfs/go-ipfs/pull/7930/commits/a35ffee1364b72e50f66fdd1a03f7a3665d99309  and added query roundtrip to ensure nothing weird ends up in HTML returned to the browser.

It now turns `?q=<script>foobar` to `?q=%3Cscript%3Efoobar`